src/UnderlineNav2/UnderlineNav.tsx

```diff
@@ -0,0 +1,161 @@
+import React, {useRef, forwardRef, useCallback, useState, MutableRefObject, RefObject} from 'react'
+import Box from '../Box'
+import {merge, SxProp, BetterSystemStyleObject} from '../sx'
+import {UnderlineNavContext} from './UnderlineNavContext'
+import {ActionMenu} from '../ActionMenu'
+import {ActionList} from '../ActionList'
+import {useResizeObserver, ResizeObserverEntry} from '../hooks/useResizeObserver'
+
+type Overflow = 'auto' | 'menu' | 'scroll'
+type ChildWidthArray = Array<{width: number}>
+type ResponsiveProps = {
+  items: Array<React.ReactElement>
+  actions: Array<React.ReactElement>
+}
+const overflowEffect = (
+  width: number,
+  childArray: Array<React.ReactElement>,
+  childWidthArray: ChildWidthArray,
+  callback: (props: ResponsiveProps) => void
+) => {
+  if (childWidthArray.length === 0) {
+    callback({items: childArray, actions: []})
+  }
+
+  // do this only for overflow
+  const numberOfItemsPossible = calculatePossibleItems(childWidthArray, width)
+  const items: Array<React.ReactElement> = []
+  const actions: Array<React.ReactElement> = []
+
+  // eslint-disable-next-line github/array-foreach
+  childArray.forEach((child: React.ReactElement, index: number) => {
+    if (index < numberOfItemsPossible) {
+      items.push(child)
+    } else {
+      actions.push(child)
+    }
+  })
+  callback({items, actions})
+}
+
+export type {ResponsiveProps}
+
+function getValidChildren(children: React.ReactNode) {
+  return React.Children.toArray(children).filter(child => React.isValidElement(child)) as React.ReactElement[]
+}
+
+function calculatePossibleItems(childWidthArray: ChildWidthArray, width: number) {
+  let breakpoint = childWidthArray.length - 1
+  let sumsOfChildWidth = 0
+  for (const [index, childWidth] of childWidthArray.entries()) {
+    if (sumsOfChildWidth > 0.5 * width) {
```

Comment:
Why did we choose 1/2 of the width as the cutoff? I think we could pick a higher threshold, or calculate based on the width of UnderlineNav minus the width of the dropdown button.

Reply:
Yeah, this calculation really didn't evolve with feedback. But I feel we could tweak it after merging the draft? I'm figuring out how to write tests for this. Then we can experiment with different cut-offs?